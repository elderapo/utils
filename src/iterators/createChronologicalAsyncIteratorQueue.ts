import { ChronologicalAsyncIteratorQueue, CreateIterator } from "./ChronologicalAsyncIteratorQueue";

export const createChronologicalAsyncIteratorQueue = <VALUE>(
  createIterators: CreateIterator<VALUE>[]
) => {
  const q = new ChronologicalAsyncIteratorQueue(createIterators);

  return q.finalize();
};
