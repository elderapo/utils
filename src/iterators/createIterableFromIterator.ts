import "core-js/modules/es7.symbol.async-iterator";

export const createIterableFromIterator = <T>(iterator: AsyncIterator<T>): AsyncIterable<T> => {
  return {
    [Symbol.asyncIterator]: () => iterator
  };
};
