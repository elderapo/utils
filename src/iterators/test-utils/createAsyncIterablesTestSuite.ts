import { sleep } from "../../timers";

export const createAsyncIterablesTestSuite = () => {
  const callResults: string[] = [];

  const createTestableAsyncIteratorThatThrowsAt = (
    from: number,
    to: number,
    delay: number,
    throwAt: number,
    identifier: string
  ) => {
    const state = {
      start: from,
      end: to,
      current: from,
      delay,
      exhaused: false
    };

    const asyncIterator: AsyncIterator<number> = {
      async next() {
        callResults.push(JSON.stringify(["next", identifier, state]));

        if (state.current >= state.end || state.exhaused) {
          return { done: true, value: state.current };
        }

        await sleep(delay);

        if (state.current === throwAt) {
          throw new Error(`REQUESTED_THROW_${identifier}_${state.current}`);
        }

        return { value: state.current++, done: false };
      },
      async throw(e) {
        callResults.push(JSON.stringify(["throw", identifier, state]));

        throw e;
      },
      async return() {
        callResults.push(JSON.stringify(["return", identifier, state]));

        state.exhaused = true;
        return { done: true, value: state.current };
      }
    };
    return asyncIterator;
  };

  const createTestableAsyncIterator = (
    from: number,
    to: number,
    delay: number,
    identifier: string
  ) =>
    createTestableAsyncIteratorThatThrowsAt(from, to, delay, (null as any) as number, identifier);

  return { createTestableAsyncIterator, createTestableAsyncIteratorThatThrowsAt, callResults };
};
