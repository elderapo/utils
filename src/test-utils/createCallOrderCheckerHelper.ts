export type CallOrderCheckerHelper = () => number;

export const createCallOrderCheckerHelper = (): CallOrderCheckerHelper => {
  let i = 0;

  return () => {
    const callID = i++;

    return callID;
  };
};
