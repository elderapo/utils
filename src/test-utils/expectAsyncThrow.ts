export const expectAsyncThrow = async <FUNCTION_ARGS extends Array<any>>(
  fn: <FUNCTION_ARGS extends Array<any>>(...args: FUNCTION_ARGS) => Promise<any> | any,
  fnArgs: FUNCTION_ARGS,
  expectedError: Error
): Promise<void> => {
  let error: Error | null = null;

  try {
    await fn(...fnArgs);
  } catch (ex) {
    error = ex;
  }

  expect(error).toEqual(expectedError);
};
