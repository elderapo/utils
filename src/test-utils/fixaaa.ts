export const fixAsyncIteratorSymbol = () => {
  // tslint:disable-next-line
  if (typeof Symbol["asyncIterator"] === "undefined") {
    (Symbol as any)["asyncIterator"] = Symbol.for("asyncIterator");
  }
};
