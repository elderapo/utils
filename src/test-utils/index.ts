export * from "./createCallOrderCheckerHelper";
export * from "./expectAsyncThrow";
export * from "./fixaaa";
