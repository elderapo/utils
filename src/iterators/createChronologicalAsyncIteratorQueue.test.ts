import { expectAsyncThrow } from "../test-utils";
import { createChronologicalAsyncIteratorQueue } from "./createChronologicalAsyncIteratorQueue";
import { createIterableFromIterator } from "./createIterableFromIterator";
import { createAsyncIterablesTestSuite } from "./test-utils";

describe("createChronologicalAsyncIteratorQueue", () => {
  it("does not accept empty createIterators array", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    expect(() => createChronologicalAsyncIteratorQueue([])).toThrow();
  });

  it("correctly works with single item", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 10, "AAA")
    ]);

    const values: number[] = [];

    for await (let a of createIterableFromIterator(combined)) {
      values.push(a);
    }

    expect(values).toMatchSnapshot();
  });

  it("chronologically yields values if first item is faster than second", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 20, "AAA"),
      () => createTestableAsyncIterator(10, 20, 23, "BBB")
    ]);

    const values: number[] = [];

    for await (let a of createIterableFromIterator(combined)) {
      values.push(a);
    }

    expect(values).toMatchSnapshot();
  });

  it("chronologically yields values if first item is slower than second", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 23, "AAA"),
      () => createTestableAsyncIterator(10, 20, 20, "BBB")
    ]);

    const values: number[] = [];

    for await (let a of createIterableFromIterator(combined)) {
      values.push(a);
    }

    expect(values).toMatchSnapshot();
  });

  it("chronologically yields values if first item is faster than second and there is a break somewhere in first item", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 20, "AAA"),
      () => createTestableAsyncIterator(10, 20, 23, "BBB")
    ]);

    const values: number[] = [];

    for await (let a of createIterableFromIterator(combined)) {
      values.push(a);

      if (a === 3) {
        break;
      }
    }

    expect(values).toMatchSnapshot();
  });

  it("chronologically yields values if first item is slower than second and there is a break somewhere in first item", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 23, "AAA"),
      () => createTestableAsyncIterator(10, 20, 20, "BBB")
    ]);

    const values: number[] = [];

    for await (let a of createIterableFromIterator(combined)) {
      values.push(a);

      if (a === 3) {
        break;
      }
    }

    expect(values).toMatchSnapshot();
  });

  it("chronologically yields values if first item is faster than second and there is a break somewhere in second item", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 20, "AAA"),
      () => createTestableAsyncIterator(10, 20, 23, "BBB")
    ]);

    const values: number[] = [];

    for await (let a of createIterableFromIterator(combined)) {
      values.push(a);

      if (a === 15) {
        break;
      }
    }

    expect(values).toMatchSnapshot();
  });

  it("chronologically yields values if first item is slower than second and there is a break somewhere in second item", async () => {
    const { createTestableAsyncIterator } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 23, "AAA"),
      () => createTestableAsyncIterator(10, 20, 20, "BBB")
    ]);

    const values: number[] = [];

    for await (let a of createIterableFromIterator(combined)) {
      values.push(a);

      if (a === 15) {
        break;
      }
    }

    expect(values).toMatchSnapshot();
  });

  it("correctly handles throw on first item(0 index)", async () => {
    const {
      createTestableAsyncIterator,
      createTestableAsyncIteratorThatThrowsAt
    } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIteratorThatThrowsAt(0, 10, 10, 0, "AAA"),
      () => createTestableAsyncIterator(10, 20, 50, "BBB")
    ]);

    const values: number[] = [];

    await expectAsyncThrow(
      async () => {
        for await (let a of createIterableFromIterator(combined)) {
          values.push(a);
        }
      },
      [],
      new Error("REQUESTED_THROW_AAA_0")
    );

    expect(values).toMatchSnapshot();
  });

  it("correctly handles throw on first item(3 index)", async () => {
    const {
      createTestableAsyncIterator,
      createTestableAsyncIteratorThatThrowsAt
    } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIteratorThatThrowsAt(0, 10, 10, 3, "AAA"),
      () => createTestableAsyncIterator(10, 20, 50, "BBB")
    ]);

    const values: number[] = [];

    await expectAsyncThrow(
      async () => {
        for await (let a of createIterableFromIterator(combined)) {
          values.push(a);
        }
      },
      [],
      new Error("REQUESTED_THROW_AAA_3")
    );

    expect(values).toMatchSnapshot();
  });

  it("correctly handles throw on second item(0 index)", async () => {
    const {
      createTestableAsyncIterator,
      createTestableAsyncIteratorThatThrowsAt,
      callResults
    } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 10, "AAA"),
      () => createTestableAsyncIteratorThatThrowsAt(10, 20, 50, 10 + 0, "BBB")
    ]);

    const values: number[] = [];

    await expectAsyncThrow(
      async () => {
        for await (let a of createIterableFromIterator(combined)) {
          values.push(a);
        }
      },
      [],
      new Error("REQUESTED_THROW_BBB_10")
    );

    expect(values).toMatchSnapshot();
  });

  it("correctly handles throw on second item(3 index)", async () => {
    const {
      createTestableAsyncIterator,
      createTestableAsyncIteratorThatThrowsAt
    } = createAsyncIterablesTestSuite();

    const combined = createChronologicalAsyncIteratorQueue([
      () => createTestableAsyncIterator(0, 10, 10, "AAA"),
      () => createTestableAsyncIteratorThatThrowsAt(10, 20, 50, 10 + 3, "BBB")
    ]);

    const values: number[] = [];

    await expectAsyncThrow(
      async () => {
        for await (let a of createIterableFromIterator(combined)) {
          values.push(a);
        }
      },
      [],
      new Error("REQUESTED_THROW_BBB_13")
    );

    expect(values).toMatchSnapshot();
  });
});
