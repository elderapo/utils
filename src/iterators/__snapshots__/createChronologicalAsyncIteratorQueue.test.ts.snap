// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createChronologicalAsyncIteratorQueue chronologically yields values if first item is faster than second 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
]
`;

exports[`createChronologicalAsyncIteratorQueue chronologically yields values if first item is faster than second and there is a break somewhere in first item 1`] = `
Array [
  0,
  1,
  2,
  3,
]
`;

exports[`createChronologicalAsyncIteratorQueue chronologically yields values if first item is faster than second and there is a break somewhere in second item 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
]
`;

exports[`createChronologicalAsyncIteratorQueue chronologically yields values if first item is slower than second 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
]
`;

exports[`createChronologicalAsyncIteratorQueue chronologically yields values if first item is slower than second and there is a break somewhere in first item 1`] = `
Array [
  0,
  1,
  2,
  3,
]
`;

exports[`createChronologicalAsyncIteratorQueue chronologically yields values if first item is slower than second and there is a break somewhere in second item 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
]
`;

exports[`createChronologicalAsyncIteratorQueue correctly handles throw on first item(0 index) 1`] = `Array []`;

exports[`createChronologicalAsyncIteratorQueue correctly handles throw on first item(3 index) 1`] = `
Array [
  0,
  1,
  2,
]
`;

exports[`createChronologicalAsyncIteratorQueue correctly handles throw on second item(0 index) 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
]
`;

exports[`createChronologicalAsyncIteratorQueue correctly handles throw on second item(3 index) 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
]
`;

exports[`createChronologicalAsyncIteratorQueue correctly works with single item 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
]
`;
